src/Jobs/BatchJob.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+namespace SilverStripe\Forager\Jobs;
+
+use SilverStripe\Core\Config\Configurable;
+use SilverStripe\Forager\Service\IndexConfiguration;
+use Symbiote\QueuedJobs\Services\AbstractQueuedJob;
+use Symbiote\QueuedJobs\Services\QueuedJob;
+
+abstract class BatchJob extends AbstractQueuedJob implements QueuedJob
+{
+
+    use Configurable;
+
+    /**
+     * A cooldown period between each batched process in the job in milliseconds
+     */
+    private static int $batch_cooldown_ms = 0;
+
+    protected function cooldown(): void
+    {
+        $cooldown = $this->config()->get('batch_cooldown_ms');
+
+        if (!$cooldown) {
+            return;
+        }
+
+        // Milliseconds to microseconds
+        $cooldown *= 1000;
+
+        usleep($cooldown);
```

Comment:
Ah yea, true. That's a much better idea. Thanks :)

Reply:
I'll prob also swap this to seconds instead of ms. I haven't looked, but I assume start after is date/time, not date/time/ms.

Reply:
just a thought we could also run the cooldown if we get a 429 response? Kinda like error handling? @chrispenny

Reply:
Having a bit of a read through the QueuedJobs module. I think the `StartAfter` is only used for the initial fetch. It then just has a `while` loop to run through the steps. We could add a new mechanism to kinda "bail out" and put the job back in the queue.

I am wondering though if blocking the queue is **what I want** out of this feature. Potentially it is. If the goal of this cooldown is to create "idle time" so that resources can recover, then I'm not sure it makes sense to allow other jobs to be picked up in the meantime to be processed.

Reply:
> just a thought we could also run the cooldown if we get a 429 response?

Does QueuedJobs have a built in "retry" mechanism, or was that something bespoke that we built in the past?

A retry with a `StartAfter` would definitely make sense.

Reply:
Just chucking a summary of our Zoom convo in here for the record (hopefully I am remembering correctly!).

We'll add a "retry" mechanism for our Jobs to this PR. The mechanism is applied to the QueuedJobs service, so the extension does apply to any/all Jobs, but we'll add configuration to target it towards Jobs specificalliy from this module. The retry mechanism will check the Exception thrown for certain status code (initially `429` and `504`), and if we haven't hit the "retry limit" we'll set the status back to "New" with a `StartAfter` set with a backoff period.

We'll make all of these settings configurable, so developers can change the filters, the backoff period, etc.

We'll keep the `sleep()` feature around. The default is `0`, and it might just be another useful lever for devs to pull if they need to slow things down.